src/sst/core/main.cc

```diff
@@ -482,39 +519,90 @@ main(int argc, char* argv[])
+        // but the same code path
+        if ( world_size.rank == 1 && world_size.thread == 1 ) cfg.partitioner = "sst.single";
 
-    try {
-        if ( partitioner->requiresConfigGraph() ) { partitioner->performPartition(graph); }
-        else {
-            PartitionGraph* pgraph;
-            if ( myRank.rank == 0 ) { pgraph = graph->getCollapsedPartitionGraph(); }
+        // Get the partitioner.  Built in partitioners are in the "sst" library.
+        SSTPartitioner* partitioner = factory->CreatePartitioner(cfg.partitioner, world_size, myRank, cfg.verbose);
+
+        try {
+            if ( partitioner->requiresConfigGraph() ) { partitioner->performPartition(graph); }
             else {
-                pgraph = new PartitionGraph();
-            }
+                PartitionGraph* pgraph;
+                if ( myRank.rank == 0 ) { pgraph = graph->getCollapsedPartitionGraph(); }
+                else {
+                    pgraph = new PartitionGraph();
+                }
 
-            if ( myRank.rank == 0 || partitioner->spawnOnAllRanks() ) {
-                partitioner->performPartition(pgraph);
+                if ( myRank.rank == 0 || partitioner->spawnOnAllRanks() ) {
+                    partitioner->performPartition(pgraph);
 
-                if ( myRank.rank == 0 ) graph->annotateRanks(pgraph);
-            }
+                    if ( myRank.rank == 0 ) graph->annotateRanks(pgraph);
+                }
 
-            delete pgraph;
+                delete pgraph;
+            }
         }
+        catch ( std::exception& e ) {
+            g_output.fatal(CALL_INFO, -1, "Error encountered during graph partitioning phase: %s\n", e.what());
+        }
+
+        delete partitioner;
     }
-    catch ( std::exception& e ) {
-        g_output.fatal(CALL_INFO, -1, "Error encountered during graph partitioning phase: %s\n", e.what());
-    }
+    else {
+        // Need to make sure all the cross rank links have matching
+        // IDs on both sides.
+
+        // We'll do this by hashing the link name and then checking
+        // for clashes.  For now, a clash will casue a fatal.  Later,
+        // we'll add a fallback.
+
+        std::set<uint32_t> clashes;
+
+        ConfigComponentMap_t& comps = graph->getComponentMap();
+        ConfigLinkMap_t&      links = graph->getLinkMap();
+        for ( ConfigLinkMap_t::iterator iter = links.begin(); iter != links.end(); ++iter ) {
+            ConfigLink& clink = *iter;
+            RankInfo    rank[2];
+            rank[0] = comps[COMPONENT_ID_MASK(clink.component[0])]->rank;
+            rank[1] = comps[COMPONENT_ID_MASK(clink.component[1])]->rank;
+            if ( rank[0].rank == rank[1].rank || (rank[0].rank != myRank.rank && rank[1].rank != myRank.rank) )
+                continue;
+
+            // Found link that goes from my rank to another one.  Get
+            // the name and generate the hash
+            const char* key  = clink.name.c_str();
```

Comment:
Any reason you don't just use `std::hash`?

Reply:
Not really.  I just used the same hashing function that we use to hash class names for serialization.  There was a reason we used a custom hash over std::hash in serialization, but I've long forgotten the reason.  There's a possibility it was just because std::hash wasn't supported for string at the time, but I decided to just use the same one instead of spending time to try to remember the reasoning.